File: android/app/src/main/java/it/curzel/bitscape/engine/GameEngine.kt
package it.curzel.bitscape.engine

import android.content.Context
import android.graphics.Bitmap
import android.graphics.RectF
import android.util.Log
import android.util.Size
import it.curzel.bitscape.analytics.RuntimeEventsBroker
import it.curzel.bitscape.AssetUtils
import it.curzel.bitscape.analytics.RuntimeEvent
import it.curzel.bitscape.controller.EmulatedKey
import it.curzel.bitscape.gamecore.AmmoRecap
import it.curzel.bitscape.gamecore.GameState
import it.curzel.bitscape.gamecore.FRect
import it.curzel.bitscape.gamecore.NativeLib
import it.curzel.bitscape.gamecore.RenderableItem
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException
import java.util.Locale

class GameEngine(
    private val context: Context,
    private val nativeLib: NativeLib,
    private val audioEngine: AudioEngine,
    private val broker: RuntimeEventsBroker,
    private val scope: CoroutineScope
) {
    private val _gameState = MutableStateFlow<GameState?>(null)
    val gameState: StateFlow<GameState?> = _gameState.asStateFlow()

    private val _weapons = MutableStateFlow<List<AmmoRecap>>(emptyList())
    val weapons: StateFlow<List<AmmoRecap>> = _weapons.asStateFlow()

    private val _isLoading = MutableStateFlow(true)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    var windowWidth: Float = 0.0f
    var windowHeight: Float = 0.0f
    var fps = 0.0

    private var isGamePaused = false
    private var currentWorldId = 0u
    private var lastFpsUpdate = System.currentTimeMillis()
    private var frameCount = 0

    var renderingScale = 1f
    var cameraViewport = FRect.zero()
    var canRender = true

    private var isLimitedVisibility: Boolean = false
    private var isNight: Boolean = false

    private val keyPressed = mutableSetOf<EmulatedKey>()
    private val keyDown = mutableSetOf<EmulatedKey>()

    private var worldHeight = 0
    private var worldWidth = 0

    private var tileMapImages = emptyList<Bitmap>()
    private var currentBiomeVariant = 0

    private val renderingScaleUseCase = RenderingScaleUseCase(context)
    private val tileMapsStorage = TileMapsStorage(context)
    private val tileSize = NativeLib.TILE_SIZE.toFloat()

    init {
        val dataPath = AssetUtils.extractAssetFolder(context, "data", "data")
        val langPath = AssetUtils.extractAssetFolder(context, "lang", "lang")

        nativeLib.initializeConfig(
            baseEntitySpeed = NativeLib.TILE_SIZE * 1.6f,
            currentLang = currentLang(),
            levelsPath = dataPath,
            speciesPath = "$dataPath/species.json",
            keyValueStoragePath = storagePath(),
            localizedStringsPath = langPath
        )
        nativeLib.initializeGame()
    }

    fun update(deltaTime: Float) {
        updateKeyboardState(deltaTime)

        val wasPaused = isGamePaused
        if (!wasPaused) {
            nativeLib.updateGame(deltaTime)
        }
        fetchRenderingInfo()
        handleWorldChanged()
        updateFpsCounter()
        flushKeyboard()

        if (!wasPaused) {
            audioEngine.updateSoundEffects()
        }

        val nextState = nativeLib.gameState()
        if (nextState.shouldPauseGame()) {
            pauseGame()
        }
        _gameState.value = nextState
        _weapons.value = nativeLib.weapons(nextState.currentPlayerIndex)
    }

    fun pauseGame() {
        isGamePaused = true
    }

    fun resumeGame() {
        isGamePaused = false
    }

    fun startNewGame() {
        resumeGame()
        broker.send(RuntimeEvent.NewGame)
        nativeLib.startNewGame()
    }

    fun renderableItems(): List<RenderableItem> {
        return nativeLib.fetchRenderableItems()
    }

    fun setupChanged(width: Float, height: Float) {
        renderingScale = renderingScaleUseCase.current()
        windowWidth = width
        windowHeight = height
        nativeLib.windowSizeChanged(width, height, renderingScale)
        worldWidth = nativeLib.currentWorldWidth()
        worldHeight = nativeLib.currentWorldHeight()
        updateTileMapImages()
    }

    fun renderingFrame(entity: RenderableItem): RectF {
        return renderingFrame(entity.frame)
    }

    fun tileMapImage(): Bitmap? {
        return tileMapImages.getOrNull(currentBiomeVariant) ?: tileMapImages.firstOrNull()
    }

    fun setKeyDown(key: EmulatedKey) {
        if (keyDown.add(key)) {
            keyPressed.add(key)
        }
    }

    fun setKeyUp(key: EmulatedKey) {
        keyPressed.remove(key)
        keyDown.remove(key)
    }

    fun fastTravelOptions(): IntArray {
        return nativeLib.fastTravelOptions()
    }

    fun cancelFastTravel() {
        nativeLib.cancelFastTravel()
        resumeGame()
    }

    fun handleFastTravel(destination: Int) {
        nativeLib.handleFastTravel(destination)
        resumeGame()
    }

    fun exitPvp() {
        nativeLib.exitPvpArena()
        resumeGame()
    }

    fun cancelPvpArena() {
        nativeLib.cancelPvpArenaRequest()
        resumeGame()
    }

    fun handlePvpArena(numberOfPlayers: Int) {
        nativeLib.handlePvpArena(numberOfPlayers)
        resumeGame()
    }

    private fun currentPlayerIndex(): Int {
        return gameState.value?.currentPlayerIndex ?: 0
    }

    private fun flushKeyboard() {
        keyPressed.clear()
        keyDown.removeAll(
            listOf(
                EmulatedKey.CLOSE_RANGE_ATTACK,
                EmulatedKey.RANGED_ATTACK,
                EmulatedKey.BACKSPACE,
                EmulatedKey.CONFIRM,
                EmulatedKey.ESCAPE,
                EmulatedKey.MENU
            )
        )
    }

    private fun updateKeyboardState(deltaTime: Float) {
        if (nativeLib.isTurnPrep()) {
            return
        }

        (0..<NativeLib.MAX_PLAYERS).forEach { playerIndex ->
            if (playerIndex == currentPlayerIndex()) {
                nativeLib.updateKeyboard(
                    player = playerIndex,
                    upPressed = keyPressed.contains(EmulatedKey.UP),
                    rightPressed = keyPressed.contains(EmulatedKey.RIGHT),
                    downPressed = keyPressed.contains(EmulatedKey.DOWN),
                    leftPressed = keyPressed.contains(EmulatedKey.LEFT),
                    upDown = keyDown.contains(EmulatedKey.UP),
                    rightDown = keyDown.contains(EmulatedKey.RIGHT),
                    downDown = keyDown.contains(EmulatedKey.DOWN),
                    leftDown = keyDown.contains(EmulatedKey.LEFT),
                    escapePressed = keyPressed.contains(EmulatedKey.ESCAPE),
                    menuPressed = keyPressed.contains(EmulatedKey.MENU),
                    confirmPressed = keyPressed.contains(EmulatedKey.CONFIRM),
                    closeAttackPressed = keyPressed.contains(EmulatedKey.CLOSE_RANGE_ATTACK),
                    rangedAttackPressed = keyPressed.contains(EmulatedKey.RANGED_ATTACK),
                    timeSinceLastUpdate = deltaTime
                )
            } else {
                nativeLib.updateKeyboard(
                    player = playerIndex,
                    upPressed = false,
                    rightPressed = false,
                    downPressed = false,
                    leftPressed = false,
                    upDown = false,
                    rightDown = false,
                    downDown = false,
                    leftDown = false,
                    escapePressed = false,
                    menuPressed = false,
                    confirmPressed = false,
                    closeAttackPressed = false,
                    rangedAttackPressed = false,
                    timeSinceLastUpdate = deltaTime
                )
            }
        }
    }

    private fun updateFpsCounter() {
        frameCount++
        val now = System.currentTimeMillis()
        val elapsed = (now - lastFpsUpdate) / 1000.0

        if (elapsed >= 1.0) {
            fps = frameCount / elapsed
            frameCount = 0
            lastFpsUpdate = now
        }
    }

    private fun handleWorldChanged() {
        val newWorld = nativeLib.currentWorldId().toUInt()
        if (newWorld == currentWorldId) { return }
        _isLoading.value = true
        canRender = false

        scope.launch {
            delay(300)
            canRender = true
            _isLoading.value = false
        }

        broker.send(RuntimeEvent.WorldTransition(currentWorldId, newWorld))
        currentWorldId = newWorld
        isNight = nativeLib.isNight()
        isLimitedVisibility = nativeLib.isLimitedVisibility()
        keyDown.clear()
        keyPressed.clear()
        updateTileMapImages()
        audioEngine.updateSoundTrack()
    }

    private fun updateTileMapImages() {
        tileMapImages = tileMapsStorage.images(currentWorldId)
    }

    private fun ensureJsonFileExists(file: File) {
        if (!file.exists()) {
            try {
                file.parentFile?.mkdirs()
                file.createNewFile()
                file.writeText("{}")
                Log.d("GameEngine", "Created new file: ${file.absolutePath}")
            } catch (e: IOException) {
                Log.e("GameEngine", "Failed to create file: ${file.absolutePath}", e)
            }
        } else {
            Log.d("GameEngine", "File already exists: ${file.absolutePath}")
        }
    }

    private fun currentLang(): String {
        return Locale.getDefault().language
            .lowercase()
            .replace("-", "_")
            .split("_")
            .firstOrNull() ?: "en"
    }

    private fun storagePath(): String {
        val fileName = "save.json"
        val file = File(context.filesDir, fileName)
        ensureJsonFileExists(file)
        return file.absolutePath
    }

    private fun renderingFrame(frame: FRect): RectF {
        val actualCol = frame.x - cameraViewport.x
        val actualRow = frame.y - cameraViewport.y

        val x = actualCol * tileSize * renderingScale
        val y = actualRow * tileSize * renderingScale

        return RectF(
            x, y,
            x + (frame.w * tileSize) * renderingScale,
            y + (frame.h * tileSize) * renderingScale
        )
    }

    fun isLimitedVisibility(): Boolean {
        return isLimitedVisibility
    }

    fun isNight(): Boolean {
        return isNight
    }

    private fun fetchRenderingInfo() {
        currentBiomeVariant = nativeLib.currentBiomeTilesVariant()
        cameraViewport = nativeLib.cameraViewport().toRect()
    }

    private fun FloatArray.toRect(): FRect {
        return FRect(this[0], this[1], this[2], this[3])
    }

    fun revive() {
        nativeLib.revive()
        resumeGame()
    }

    fun isPvp(): Boolean {
        return nativeLib.isPvp()
    }

    fun setWeaponEquippedForCurrentPlayer(weaponSpeciesId: Int) {
        nativeLib.setWeaponEquipped(weaponSpeciesId, currentPlayerIndex())
    }
}

File: ios/SneakBit/Engine/GameEngine.swift
import Foundation
import Combine
import UIKit
import Schwifty

class GameEngine {
    @Inject private var audioEngine: AudioEngine
    @Inject private var broker: RuntimeEventsBroker
    @Inject private var renderingScaleUseCase: RenderingScaleUseCase
    @Inject private var tileMapsStorage: TileMapsStorage
    
    private var currentState: GameState? = nil {
        didSet {
            _state.send(currentState)
        }
    }
    
    private let _state = CurrentValueSubject<GameState?, Never>(nil)
    let isLoading = CurrentValueSubject<Bool, Never>(true)
    
    var size: CGSize = .zero
    var fps: Double = 0.0
    
    var isLandscape: Bool {
        cameraViewport.w >= cameraViewport.h
    }
    
    var currentPlayerIndex: UInt {
        currentState?.current_player_index ?? 0
    }
    
    private var currentWorldId: UInt32 = 0
    private var lastFpsUpdate: Date = Date()
    private var frameCount: Int = 0
    
    let tileSize = CGFloat(TILE_SIZE)
    private(set) var renderingScale: CGFloat = 1
    private(set) var cameraViewport: IntRect = .zero
    private(set) var cameraViewportOffset: Vector2d = .zero
    private(set) var safeAreaInsets: UIEdgeInsets = .zero
    private(set) var canRender: Bool = true
    private(set) var isNight: Bool = false
    private(set) var isLimitedVisibility: Bool = false
        
    private var keyPressed = Set<EmulatedKey>()
    private var keyDown = Set<EmulatedKey>()
    
    private var worldHeight: Int = 0
    private var worldWidth: Int = 0
    private var isGamePaused: Bool = false
    
    private(set) var tileMapImages: [UIImage] = []
    private var currentBiomeVariant: Int = 0
    
    init() {
        initialize_config(
            true,
            Float(TILE_SIZE * 1.8),
            currentLang(),
            dataFolder(),
            speciesJson(),
            saveJson(),
            langFolder()
        )
        initialize_game(GameMode_RealTimeCoOp)
    }
    
    func gameState() -> AnyPublisher<GameState, Never> {
        _state
            .compactMap { $0 }
            .receive(on: DispatchQueue.main)
            .eraseToAnyPublisher()
    }
    
    func update(deltaTime: Float) {
        updateKeyboardState(timeSinceLastUpdate: deltaTime)
        
        let wasPaused = isGamePaused
        if !wasPaused {
            update_game(deltaTime)
        }
        
        fetchRenderingInfo()
        handleWorldChanged()
        updateFpsCounter()
        flushKeyboard()
        
        if !wasPaused {
            audioEngine.updateSoundEffects()
        }
        
        let newState = game_state()
        if newState.shouldPauseGame() {
            pauseGame()
        }
        currentState = newState
    }
    
    private func handleWorldChanged() {
        let newWorld = current_world_id()
        guard newWorld != currentWorldId else { return }
        isLoading.send(true)
        canRender = false
        DispatchQueue.main.asyncAfter(deadline: .now() + WORLD_TRANSITION_TIME) {
            self.canRender = true
            self.isLoading.send(false)
        }
        
        broker.send(.worldTransition(source: currentWorldId, destination: current_world_id()))
        currentWorldId = newWorld
        isNight = is_night()
        isLimitedVisibility = is_limited_visibility()
        keyDown.removeAll()
        keyPressed.removeAll()
        updateTileMapImages()
        audioEngine.updateSoundTrack()
    }
    
    func startNewGame() {
        resumeGame()
        broker.send(.newGame)
        start_new_game()
    }
    
    func pauseGame() {
        isGamePaused = true
    }
    
    func resumeGame() {
        isGamePaused = false
    }
    
    func renderEntities(_ render: @escaping (RenderableItem) -> Void) {
        fetchRenderableItems { items in
            items.forEach { item in
                render(item)
            }
        }
    }

    func setupChanged(safeArea: UIEdgeInsets?, windowSize: CGSize, screenScale: CGFloat?) {
        safeArea.let { safeAreaInsets = $0 }
        renderingScale = renderingScaleUseCase.calculate(windowSize: windowSize, screenScale: screenScale)
        size = windowSize
        window_size_changed(Float(size.width), Float(size.height), Float(renderingScale))
        worldHeight = Int(current_world_height())
        worldWidth = Int(current_world_width())
    }
    
    func renderingFrame(for entity: RenderableItem) -> CGRect {
        renderingFrame(for: entity.frame, offset: entity.offset)
    }
    
    func tileMapImage() -> UIImage? {
        if currentBiomeVariant < tileMapImages.count {
            tileMapImages[currentBiomeVariant]
        } else {
            nil
        }
    }
    
    private func renderingFrame(for frame: IntRect, offset: Vector2d = .zero) -> CGRect {
        let actualCol = CGFloat(frame.x - cameraViewport.x)
        let actualOffsetX = CGFloat(offset.x - cameraViewportOffset.x)

        let actualRow = CGFloat(frame.y - cameraViewport.y)
        let actualOffsetY = CGFloat(offset.y - cameraViewportOffset.y)
        
        return CGRect(
            x: (actualCol * tileSize + actualOffsetX) * renderingScale,
            y: (actualRow * tileSize + actualOffsetY) * renderingScale,
            width: CGFloat(frame.w) * tileSize * renderingScale,
            height: CGFloat(frame.h) * tileSize * renderingScale
        )
    }
    
    func setKeyDown(_ key: EmulatedKey) {
        if keyDown.contains(key) {
            return
        }
        keyPressed.insert(key)
        keyDown.insert(key)
    }
    
    func setKeyUp(_ key: EmulatedKey) {
        keyPressed.remove(key)
        keyDown.remove(key)
    }
    
    private func flushKeyboard() {
        for key in keyPressed where key.isMovement {
            keyDown.insert(key)
        }
        keyPressed.removeAll()
        
        keyDown.remove(.closeRangeAttack)
        keyDown.remove(.rangedAttack)
        keyDown.remove(.backspace)
        keyDown.remove(.confirm)
        keyDown.remove(.escape)
        keyDown.remove(.menu)
    }
    
    private func updateKeyboardState(timeSinceLastUpdate: Float) {
        guard !is_turn_prep() else {
            return
        }
        
        for playerIndex in 0..<MAX_PLAYERS {
            if playerIndex == currentPlayerIndex {
                update_keyboard(
                    UInt(playerIndex),
                    keyPressed.contains(.up),
                    keyPressed.contains(.right),
                    keyPressed.contains(.down),
                    keyPressed.contains(.left),
                    keyDown.contains(.up),
                    keyDown.contains(.right),
                    keyDown.contains(.down),
                    keyDown.contains(.left),
                    keyPressed.contains(.escape),
                    keyPressed.contains(.menu),
                    keyPressed.contains(.confirm),
                    keyPressed.contains(.closeRangeAttack),
                    keyPressed.contains(.rangedAttack),
                    keyPressed.contains(.backspace),
                    false,
                    timeSinceLastUpdate
                )
            } else {
                update_keyboard(
                    UInt(playerIndex),
                    false, false, false, false,
                    false, false, false, false,
                    false, false, false,
                    false, false,
                    false,
                    false,
                    timeSinceLastUpdate
                )
            }
        }
    }
    
    private func updateTileMapImages() {
        tileMapImages = tileMapsStorage.images(forWorld: currentWorldId)
    }
    
    private func updateFpsCounter() {
        frameCount += 1
        let now = Date()
        let elapsed = now.timeIntervalSince(lastFpsUpdate)
        
        if elapsed >= 1.0 {
            fps = Double(frameCount) / elapsed
            frameCount = 0
            lastFpsUpdate = now
        }
    }
    
    private func fetchRenderingInfo() {
        currentBiomeVariant = Int(current_biome_tiles_variant())
        cameraViewport = camera_viewport()
        cameraViewportOffset = camera_viewport_offset()
    }
}
